hypothesis-python/src/hypothesis/extra/numpy.py

```diff
@@ -117,6 +118,33 @@ def __init__(self, element_strategy, shape, dtype, fill, unique):
         self.element_strategy = element_strategy
         self.unique = unique
 
+        # Used by self.insert_element to check that the value can be stored
+        # in the array without e.g. overflowing.  See issue #1385.
+        if dtype.kind in (u'i', u'u'):
+            self.check_cast = lambda x: np.can_cast(x, self.dtype, 'safe')
+        elif dtype.kind == u'f' and dtype.itemsize == 2:
+            max_f2 = (2. - 2 ** -10) * 2 ** 15
+            self.check_cast = lambda x: \
+                (not np.isfinite(x)) or (-max_f2 <= x <= max_f2)
+        elif dtype.kind == u'f' and dtype.itemsize == 4:
+            max_f4 = (2. - 2 ** -23) * 2 ** 127
+            self.check_cast = lambda x: \
+                (not np.isfinite(x)) or (-max_f4 <= x <= max_f4)
+        else:
+            self.check_cast = lambda x: True
+
+    def insert_element(self, data, result, idx, strategy=None):
+        strategy = strategy or self.element_strategy
+        val = data.draw(strategy)
+        if not self.check_cast(val):
```

Comment:
We're going to be emitting a *lot* of these, and because they include the actual values (which is good!) they won't show up for duplication. I wonder if it would be worth maintaining state so as to only emit one per test case?